jenkins/scripts/test_rerun.py

```
@@ -16,6 +16,41 @@ def parse_name(classname, name, filename):
+def extract_passed_tests(output_file, xml_filenames):
+    # Walk one or more JUnit XML files and emit the names of test cases that
+    # passed (no <failure> / <error> / <skipped> child). Used by
+    # reusePassedTestResults in L0_Test.groovy to recover passes from a prior
+    # retry attempt's tarball when the build's infra-failure retry fires.
+    seen = set()
+    with open(output_file, 'w') as out_file:
+        for xml_filename in xml_filenames:
+            if not os.path.exists(xml_filename):
+                continue
+            try:
+                root = ET.parse(xml_filename).getroot()
```

Ruff (0.15.12)

[error] Using `xml` to parse untrusted data is known to be vulnerable to XML attacks; use `defusedxml` equivalents

(S314)